File: src/cpp-leetcode/first/Code_023_MergeKSortedLists.cpp
#include <vector>

using std::vector;

#include "../List.h"

/**
 * merge k sorted linked-list
 *
 * @since 2020-9-17 Thursday 11:40 - 12:00
 */
class Solution {
   private:
    ListNode* mergeTwoLists(ListNode* h1, ListNode* h2) {
        ListNode *dummy = new ListNode(-1), *cur = dummy;
        while (h1 != nullptr && h2 != nullptr) {
            if (h1->val < h2->val) {
                cur->next = h1;
                h1 = h1->next;
            } else {
                cur->next = h2;
                h2 = h2->next;
            }
            cur = cur->next;
        }
        if (h1 == nullptr) {
            cur->next = h2;
        } else {
            cur->next = h1;
        }
        return dummy->next;
    }

    ListNode* merge(vector<ListNode*> lists, int l, int r) {
        if (l == r) {
            return lists[l];
        }
        int mid = l + (r - l) / 2;
        ListNode* left = merge(lists, l, mid);
        ListNode* right = merge(lists, mid + 1, r);
        return mergeTwoLists(left, right);
    }

   public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        int size = lists.size();
        if (size == 0) {
            return nullptr;
        }
        return merge(lists, 0, size - 1);
    }
};

int main(int argc, char const* argv[]) {
    vector<int> v1 = {1, 3, 5, 7};
    vector<int> v2 = {2, 5, 8, 9};
    vector<int> v3 = {1, 10};
    ListNode* h1 = List::linkedList(v1);
    ListNode* h2 = List::linkedList(v2);
    ListNode* h3 = List::linkedList(v3);
    Solution s;
    // vector<ListNode*> lists = {h1, h2, h3};
    vector<ListNode*> lists = {};
    ListNode* res = s.mergeKLists(lists);
    List::printList(res);
}

File: src/leetcode/fourth/Code_023_MergeKSortedLists.java
package leetcode.fourth;

import leetcode.tool.LinkedListTool;
import leetcode.tool.ListNode;

/**
 * 合并k个有序链表
 *
 * @since 2020-5-6 Wednesday 16:57 - 17:22
 */
public class Code_023_MergeKSortedLists {
    ListNode mergeKLists(ListNode[] lists) {
        int len = lists.length;
        if (len == 0) return null;
        return helper(lists, 0, len - 1);
    }

    ListNode helper(ListNode[] lists, int l, int r) {
        if (l == r) return lists[l];
        int mid = l + (r - l) / 2;
        ListNode l1 = helper(lists, l, mid);
        ListNode l2 = helper(lists, mid + 1, r);
        return merge2Lists(l1, l2);
    }

    private ListNode merge2Lists(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode(Integer.MIN_VALUE), cur = dummy;
        while (l1 != null && l2 != null) {
            if (l1.val < l2.val) {
                cur.next = l1;
                l1 = l1.next;
            } else {
                cur.next = l2;
                l2 = l2.next;
            }
            cur = cur.next;
        }
        if (l1 == null) {
            cur.next = l2;
        } else {
            cur.next = l1;
        }
        return dummy.next;
    }

    public static void main(String... args) {
        Code_023_MergeKSortedLists c = new Code_023_MergeKSortedLists();
        ListNode h1 = LinkedListTool.generateList(new int[]{1, 2, 3, 4});
        ListNode h2 = LinkedListTool.generateList(new int[]{2, 5, 9});
        ListNode h3 = LinkedListTool.generateList(new int[]{1, 6, 8, 10});
        LinkedListTool.printList(c.mergeKLists(new ListNode[]{h1, h2, h3}));
        // LinkedListTool.printList(c.merge2Lists(h2, h3));
    }
}
